File: nikkake_flutter/lib/data/offline_queue.dart
import 'dart:convert';

import 'graphql_client.dart';
import 'local_db.dart';

/// オフライン記録のキュー。
///
/// 記録だけがオフライン可（ルーティンの作成・編集はオンライン必須）。
/// id は clientMutationId としてそのままサーバへ送るので、
/// 再送しても mutation_receipts とクライアント生成UUIDの主キー衝突で吸収される。
class QueuedOp {
  const QueuedOp({
    required this.id,
    required this.payload,
    required this.attempts,
    required this.nextAttemptAt,
    this.lastError,
  });

  final String id;
  final Map<String, dynamic> payload;
  final int attempts;
  final DateTime nextAttemptAt;
  final String? lastError;

  Map<String, dynamic> toJson() => {
        'id': id,
        'payload': payload,
        'attempts': attempts,
        'nextAttemptAt': nextAttemptAt.toIso8601String(),
        'lastError': lastError,
      };

  static QueuedOp fromJson(Map<String, dynamic> json) => QueuedOp(
        id: json['id'] as String,
        payload: (json['payload'] as Map).cast<String, dynamic>(),
        attempts: json['attempts'] as int? ?? 0,
        nextAttemptAt:
            DateTime.tryParse(json['nextAttemptAt'] as String? ?? '') ?? DateTime.now(),
        lastError: json['lastError'] as String?,
      );
}

class OfflineQueue {
  OfflineQueue(this.db);

  final LocalDb db;

  static const _maxBackoff = Duration(minutes: 5);

  List<QueuedOp> _read({required bool isDead}) {
    final meta = db.getMeta();
    final raw = isDead ? meta.offlineQueueDead : meta.offlineQueue;
    if (raw == null || raw.isEmpty) return [];

    try {
      return (jsonDecode(raw) as List)
          .cast<Map<String, dynamic>>()
          .map(QueuedOp.fromJson)
          .toList();
    } catch (_) {
      // 壊れたJSONで記録できなくなるのが最悪なので、握りつぶして空で復旧する
      return [];
    }
  }

  Future<void> _write({required bool isDead, required List<QueuedOp> ops}) {
    final encoded = jsonEncode(ops.map((o) => o.toJson()).toList());
    return isDead
        ? db.setMeta(offlineQueueDead: encoded)
        : db.setMeta(offlineQueue: encoded);
  }

  List<QueuedOp> pending() => _read(isDead: false);

  /// 送信できなかったもの。設定画面に出して手動対応させる
  List<QueuedOp> dead() => _read(isDead: true);

  Future<void> enqueue(String id, Map<String, dynamic> payload) async {
    final ops = pending();
    if (ops.any((o) => o.id == id)) return;

    ops.add(QueuedOp(
      id: id,
      payload: payload,
      attempts: 0,
      nextAttemptAt: DateTime.now(),
    ));
    await _write(isDead: false, ops: ops);
  }

  /// 直列に送る。失敗したら以降は次回に回して順序を守る。
  /// オフライン記録は少数なので並列化する必要がない。
  Future<int> flush(Future<void> Function(QueuedOp) send) async {
    var sent = 0;

    for (final op in pending()) {
      if (op.nextAttemptAt.isAfter(DateTime.now())) continue;

      try {
        await send(op);
        await _remove(op.id);
        sent++;
      } on PermanentFailure catch (e) {
        await _moveToDead(op, e.message);
        break;
      } catch (e) {
        await _backoff(op, '$e');
        break;
      }
    }

    return sent;
  }

  Future<void> _remove(String id) =>
      _write(isDead: false, ops: pending().where((o) => o.id != id).toList());

  Future<void> _moveToDead(QueuedOp op, String message) async {
    await _remove(op.id);
    await _write(isDead: true, ops: [
      ...dead(),
      QueuedOp(
        id: op.id,
        payload: op.payload,
        attempts: op.attempts,
        nextAttemptAt: op.nextAttemptAt,
        lastError: message,
      ),
    ]);
  }

  Future<void> _backoff(QueuedOp op, String message) async {
    final attempts = op.attempts + 1;
    var delay = Duration(seconds: 1 << op.attempts);
    if (delay > _maxBackoff) delay = _maxBackoff;

    await _write(
      isDead: false,
      ops: pending()
          .map((o) => o.id == op.id
              ? QueuedOp(
                  id: o.id,
                  payload: o.payload,
                  attempts: attempts,
                  nextAttemptAt: DateTime.now().add(delay),
                  lastError: message,
                )
              : o)
          .toList(),
    );
  }
}

File: nikkake_kmp/shared/src/commonMain/kotlin/com/myapplication/common/data/OfflineQueue.kt
package com.myapplication.common.data

import com.myapplication.common.domain.nowIso
import kotlinx.datetime.Clock
import kotlinx.datetime.Instant
import kotlinx.serialization.KSerializer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.PrimitiveKind
import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlin.time.Duration.Companion.seconds

/**
 * オフライン記録のキュー。
 *
 * 記録だけがオフライン可（ルーティンの作成・編集はオンライン必須）。
 * id は clientMutationId としてそのままサーバへ送るので、
 * 再送しても mutation_receipts とクライアント生成UUIDの主キー衝突で吸収される。
 */
@Serializable
data class QueuedOp(
    val id: String,
    val payload: JsonObject,
    val attempts: Int = 0,
    @SerialName("next_attempt_at") val nextAttemptAt: String,
    @SerialName("last_error") val lastError: String? = null,
)

object InstantAsStringSerializer : KSerializer<Instant> {
    override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor("Instant", PrimitiveKind.STRING)
    override fun serialize(encoder: Encoder, value: Instant) = encoder.encodeString(value.toString())
    override fun deserialize(decoder: Decoder): Instant = Instant.parse(decoder.decodeString())
}

class OfflineQueue(private val db: LocalDb) {

    private val json = Json { ignoreUnknownKeys = true }
    private val listSerializer = kotlinx.serialization.builtins.ListSerializer(QueuedOp.serializer())

    companion object {
        private val MAX_BACKOFF_SECONDS = (5 * 60)
    }

    private fun read(dead: Boolean): List<QueuedOp> {
        val meta = db.getMeta()
        val raw = if (dead) meta.offlineQueueDead else meta.offlineQueue
        if (raw.isNullOrEmpty()) return emptyList()

        // 壊れたJSONで記録できなくなるのが最悪なので、握りつぶして空で復旧する
        return runCatching { json.decodeFromString(listSerializer, raw) }.getOrElse { emptyList() }
    }

    private fun write(dead: Boolean, ops: List<QueuedOp>) {
        val encoded = json.encodeToString(listSerializer, ops)
        if (dead) db.setMeta(offlineQueueDead = encoded) else db.setMeta(offlineQueue = encoded)
    }

    fun pending(): List<QueuedOp> = read(dead = false)

    /** 送信できなかったもの。設定画面に出して手動対応させる */
    fun dead(): List<QueuedOp> = read(dead = true)

    fun enqueue(id: String, payload: JsonObject) {
        val ops = pending()
        if (ops.any { it.id == id }) return

        write(dead = false, ops = ops + QueuedOp(id = id, payload = payload, nextAttemptAt = nowIso()))
    }

    /**
     * 直列に送る。失敗したら以降は次回に回して順序を守る。
     * オフライン記録は少数なので並列化する必要がない。
     */
    suspend fun flush(send: suspend (QueuedOp) -> Unit): Int {
        var sent = 0

        for (op in pending()) {
            val nextAttempt = runCatching { Instant.parse(op.nextAttemptAt) }.getOrNull()
            if (nextAttempt != null && nextAttempt > Clock.System.now()) continue

            try {
                send(op)
                remove(op.id)
                sent++
            } catch (e: PermanentFailure) {
                moveToDead(op, e.message ?: "エラー")
                break
            } catch (e: Exception) {
                backoff(op, e.message ?: "エラー")
                break
            }
        }

        return sent
    }

    private fun remove(id: String) {
        write(dead = false, ops = pending().filter { it.id != id })
    }

    private fun moveToDead(op: QueuedOp, message: String) {
        remove(op.id)
        write(dead = true, ops = dead() + op.copy(lastError = message))
    }

    private fun backoff(op: QueuedOp, message: String) {
        val attempts = op.attempts + 1
        val delaySeconds = minOf(MAX_BACKOFF_SECONDS, 1 shl op.attempts)
        val next = Clock.System.now() + delaySeconds.seconds

        write(
            dead = false,
            ops = pending().map {
                if (it.id == op.id) {
                    it.copy(attempts = attempts, nextAttemptAt = next.toString(), lastError = message)
                } else it
            },
        )
    }
}
